Написание рекомендации

YaKokin opened the pull request.

YaKokin committed:
функционал выдачи рекомендации пользователю
- поправил логику сохранения рекомендации. теперь если у ресивера не было скилла, предлагаемого в рекомендации - он также добавляется в скиллы, которыми юзер теперь владеет
- провел рефакторинг
- вынес дублирующийся код с обработкой коллекций в CollectionUtils
- добавил универсальный метод для обработки коллекции с применением фильтров
(без тестов)

YaKokin committed:
функционал удаления рекомендации
(без тестов)

YaKokin committed:
функционал обновления рекомендации (не доделанный)
- добавил глобал хандлер для исключений
- добавил исключение для кейса, когда отсутствуют ресурсы
- добавил класс для респонса в случае ошибки
(без тестов)

YaKokin committed:
функционал обновления рекомендации
(без тестов)

YaKokin committed:
функционал получения рекомендации пользователя
(без тестов)

YaKokin committed:
функционал получения выданных рекомендаций пользователя
(без тестов)

YaKokin committed:
- Добавил аннотацию Positive на id принимаемый в контроллере
 - провел минорный рефакторинг, поправил код стайл

YaKokin committed:
добавил версионирование
провел минорный рефакторинг

YaKokin committed:
добавил логирование

The-Naku commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,238 @@
+    @Transactional
+    public void delete(long recommendationId) {
+        log.info("Deleting recommendation with id: {}", recommendationId);
+        Recommendation recommendation = recommendationRepo.findById(recommendationId)
```
не хочешь ли ты вынести эту строчку в отдельный метод  getRecById,, чтобы не было повторений со строкой 87 и не нагружать метод выбросом исключений. Ну и еще чтобы, если понадобится другой человек мог вызвать уже готовый метод

YaKokin replied:
1. я думаю, что выносить одну строчку в целый метод для внутреннего использования не стоит.
2. насчет внешнего вызова - пока такой необходимости нет, поэтому добавлять просто так смысла нет.
3. исключение здесь только прокидывается. обработка производится в глобал ханделере

The-Naku commented:
а еще исправь , чтобы merge был в ветку Wyvern-master-stream7. надо от нее создавать ветки с фичами и мерджить туда

YaKokin committed:
- провел рефакторинг RecommendationService
- поработал над неймингом
- провел декомпозицию методов на логические фрагменты

YaKokin committed:
- провел рефакторинг RecommendationService
- поработал над неймингом
- провел декомпозицию методов на логические фрагменты

YaKokin committed:
Merge remote-tracking branch 'origin/feature-BJS2-40459' into feature-BJS2-40459

YaKokin committed:
вынес сообщение об ошибке валидации в константу

ipasynkov-epam reviewed (approved):
Отлично
